reactor-netty-core/src/test/java/reactor/netty/ConnectionObserverContextTest.java

```diff
@@ -0,0 +1,182 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package reactor.netty;
+
+import org.junit.After;
+import org.junit.AfterClass;
+import org.junit.Before;
+import org.junit.Test;
+import org.reactivestreams.Publisher;
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+import reactor.core.CoreSubscriber;
+import reactor.core.Fuseable;
+import reactor.core.publisher.Hooks;
+import reactor.core.publisher.Operators;
+import reactor.netty.tcp.TcpClient;
+import reactor.netty.tcp.TcpServer;
+import reactor.util.context.Context;
+
+import java.util.Optional;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Function;
+import java.util.function.UnaryOperator;
+
+import static java.util.Objects.requireNonNull;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertTrue;
+
+/**
+ * This test class verifies that a stream's {@link Context} is propagated to
+ * {@link ConnectionObserver} in {@link TcpClient#doOnChannelInit}.
+ *
+ * The test is specifically interested in verifying that the available
+ * {@link Context} is propagated to {@link ConnectionObserver}, not really
+ * interested in verifying that {@link Context} is propagated from outer stream
+ * to the inner stream.  That's why it uses {@link Hooks} to propagate
+ * {@link Context} from outer stream to the inner stream.
+ */
+public class ConnectionObserverContextTest {
+
+	private static final String CONTEXT_KEY = "marcels-key";
+	private static final String CONTEXT_VALUE = "marcels-context";
+	private static final ThreadLocal<String> helloWorld = new ThreadLocal();
+
+	private static final DisposableServer server = TcpServer.create().port(0).bindNow();
+
+	@Before
+	public void before() {
+		helloWorld.set(CONTEXT_VALUE);
+		Hooks.onLastOperator(HelloWorldPropagatorSubscriber.class.getName(), HelloWorldPropagatorSubscriber.asOperator());
+	}
+
+	@After
+	public void after() {
+		helloWorld.remove();
+		Hooks.resetOnLastOperator(HelloWorldPropagatorSubscriber.class.getName());
+	}
+
+	@AfterClass
+	public static void afterClass() {
+		server.disposeNow();
+	}
+
+	@Test
+	public void testContextIsPropagatedToConnectionObserver() throws Exception {
+
+		final AtomicReference<String> contextualData = new AtomicReference<>();
+		final CountDownLatch channelInitialized = new CountDownLatch(1);
+
+		TcpClient.create()
+				.doOnChannelInit((connectionObserver, channel, remoteAddress) -> {
+					if (connectionObserver.currentContext().hasKey(CONTEXT_KEY)) {
+						contextualData.set(connectionObserver.currentContext().get(CONTEXT_KEY));
+					}
+					channelInitialized.countDown();
+				})
+				.remoteAddress(server::address)
+				.wiretap(true)
+				.connect()
+				.contextWrite(Context.of(CONTEXT_KEY, CONTEXT_VALUE))
```

Comment:
I don't understand the purpose of having both `contextWrite` and the threadLocal-to-`Context` operator decorator / hook. isn't the purpose of the change to validate that the `Context` from the `connect()` `Mono` is visible within `ConnextionObserver`? Having the value set by both an explicit operator and an operator hook makes it hard to reason about this test. If we expect one to take precedence over the other, then at a minimum the values put in the context should be different ?

Reply:
I agree with you and will remove this line.  However, the original goal was to add a context and the outer stream and the context should be available in `doOnChannelInit`.  The reason Hooks are added is because of the first issue mentioned in https://github.com/reactor/reactor-netty/issues/1327#issuecomment-707849473.  Once https://github.com/reactor/reactor-pool/issues/103 is addressed, we should hopefully not need the `Hooks`.  

For now, I will change the test to just rely on `Hooks` and will add comment why that's the case.